src/topology/mod.rs

```
@@ -1111,32 +929,65 @@ where
     }
 }
 
-impl<'a, C> AsCommunicator for Process<'a, C>
-where
-    C: 'a + Communicator,
-{
-    type Out = C;
-    fn as_communicator(&self) -> &Self::Out {
+unsafe impl<'a> AsRaw for Process<'a> {
+    type Raw = MPI_Comm;
+
+    fn as_raw(&self) -> Self::Raw {
+        return self.comm.as_raw();
+    }
+}
+
+impl<'a> sealed::AsHandle for Process<'a> {
+    fn as_handle(&self) -> &sealed::CommunicatorHandle {
         self.comm
     }
 }
 
+impl<'a> Communicator for Process<'a> {
+    fn target_size(&self) -> Rank {
+        self.size()
+    }
+}
+
+impl<'a> AsCommunicator for Process<'a> {
+    type Out = Self;
+
+    fn as_communicator(&self) -> &Self::Out {
+        self
+    }
+}
+
```

This turns out to be a bug, in that `process.as_communicator().rank()` now returns the rank of `process`, not the rank of the current process on the communicator. One can call `Communicator::rank(&process)`, but I think `process.as_communicator()` should behave like a communicator, not silently use the wrong `rank()` method.

Do you have thoughts about how to fix it? We could put an actual communicator back into `Process`? It could be `comm: &'a dyn Communicator` to avoid making it generic over `Communicator` or any other trivial wrapper around a `CommunicatorHandle`. `AnyProcess<'a>` could be abused for that purpose, but it seems decidedly unclean.

---

I am pretty convinced that the PR didn't change any behavior regarding to this.
Previously `Process` looked like this:
```rs
pub struct Process<'a, C>
where
    C: 'a + Communicator,
{
    comm: &'a C,
    rank: Rank,
}
```
and the `AsCommunicator` trait was implemented to return `comm`. Now it returns `self`, because `comm` is no longer a `Communicator` but a `CommunicatorHandle`. Instead, `Process` implements `Communicator` itself, which in turn delegates all calls to the internal `CommunicatorHandle`. So if this bug exists, I am pretty sure it existed before this PR (I haven't tested this though, so maybe I am missing something).

Not sure what even causes this, because I am not sure how MPI differentiates between a Process handle and a Communicator handle. Is the differentiation only done by choosing different rank functions?

If so, maybe we could change the `Communicator` trait impl of Process by explicitly changing its rank implementation:

```rs
impl<'a> Communicator for Process<'a> {
    // ...

    fn rank(&self) -> Rank {
        Process::rank(&self)
    }
}
```

---

Thanks for your reply and sorry I'm slow following up. The issue is that `impl Process` has a method called `rank()` referring to the rank of the designated process, while the `rank()` method on the `Communicator` trait returns the rank of self.

In the old code, `process.as_communicator()` was just a `Communicator` (no longer a `Process`) so there was only one `rank()` method. After your change here, it returned a `Process` (which implements `Communicator`, but prefers the method on `Process` instead of the trait method).

I resolved the issue in https://github.com/rsmpi/rsmpi/pull/177/, though I'm not entirely happy with the solution. Part of me wants to remove/rename the `Process::rank()` method. There are also experimental systems in which a rank is not a process, so I'm not wild about the `Process` name.

---

Oh, I see the problem now. Would it be confusing for users if the rank methods on both traits simply had different names (e.g. `communicator_rank()` and `processor_rank()`)? That way, it'd be clear which rank is called at any given point.

---

Yeah, though `comm.comm_rank()` and `process.process_rank()` are somewhat tedious and I'm not sure it's better than the current use of `AnyProcess`, at least until we get around to making `Rank` stronger (e.g., a newtype instead of an alias to `c_int`). It's necessary to have a type that can be put in an array and sorted, though it would be nice for it to retain a branding associating it with the communicator on which it's valid.

I think the current `Process` is pure ergonomics and I have mixed feelings about its existence -- many/most real-world patterns will be calling `comm.process_at_rank()` and using it immediately in exactly one point-to-point operation, but then we need run-time assertions like you'll see in `send_receive_with_tags` to check that source and destination communicators match. In a sense, it is more error-prone (weaker compile-time guarantees) to use `source_process, dest_process` than `comm, source_rank, dest_rank`.